src/pstl_offload.cpp

```diff
@@ -96,59 +97,311 @@ __original_msize(void* __user_ptr)
+    {
+        assert(__is_ptr_page_aligned(__ptr));
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        [[maybe_unused]] auto __ret = _M_map->emplace(__ptr, __ptr_desc{std::move(__device_ptr), __size});
+        assert(__ret.second); // the pointer must be unique
+    }
+
+    // nullopt means "it's not our pointer"
+    static std::optional<__ptr_desc>
+    __unregister_ptr(__large_aligned_ptrs_map& __this, void* __ptr)
+    {
+        // only page-aligned can be registered, so they may not be in the map
+        if (!__is_ptr_page_aligned(__ptr))
+        {
+            return std::nullopt;
+        }
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        if (!__this._M_map)
+        {
+            // ctor of static object not yet run, so it can't be our pointer
+            return std::nullopt;
+        }
+        auto __iter = __this._M_map->find(__ptr);
+        if (__iter == __this._M_map->end())
+        {
+            return std::nullopt;
+        }
+        __ptr_desc __header = std::move(__iter->second);
+        __this._M_map->erase(__iter);
+        return __header;
+    }
+
+    // nullopt means "it's not our pointer"
+    static std::optional<std::size_t>
+    __get_size(__large_aligned_ptrs_map& __this, void* __ptr)
+    {
+        // only page-aligned can be registered
+        if (!__is_ptr_page_aligned(__ptr))
+        {
+            return std::nullopt;
+        }
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        if (!__this._M_map)
+        {
+            // ctor of static object not yet run, so it can't be our pointer
+            return std::nullopt;
+        }
+        auto __iter = __this._M_map->find(__ptr);
+        if (__iter == __this._M_map->end())
+        {
+            return std::nullopt;
+        }
+        return __iter->second._M_requested_number_of_bytes;
+    }
+};
+
+static __large_aligned_ptrs_map __large_aligned_ptrs;
+
+void
+__block_header::__free()
+{
+    _M_uniq_const = 0;
+    sycl::context __context = _M_device.__get_context();
+    void* __original_pointer = _M_original_pointer;
+    this->~__block_header();
+    sycl::free(__original_pointer, __context);
+}
+
 static void
 __internal_free(void* __user_ptr)
 {
     if (__user_ptr != nullptr)
     {
-        __block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+        if (__block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+            __same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
+        {
+            __header->__free();
+            return;
+        }
 
-        if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
+        if (std::optional<__large_aligned_ptrs_map::__ptr_desc>
+                __desc = __large_aligned_ptrs_map::__unregister_ptr(__large_aligned_ptrs, __user_ptr))
```

Comment:
As far as I understand, the `__unregister_ptr` have try semantics and may fail. May be it makes more sense to rename it to `__try_unregister_ptr` to better explain the semantics?
The same applies to `__get_size`.

Reply:
Interesting. Is "try semantics" kind of standard terminology? I agree, the names are not clean, can it be `__unregister_known_ptr` etc?